core/src/main/java/org/apache/calcite/util/TimeString.java

```diff
@@ -76,35 +76,54 @@ private static String hms(int h, int m, int s) {
     return b.toString();
   }
 
-  /** Sets the fraction field of a {@code TimeString} to a given number
-   * of milliseconds. Nukes the value set via {@link #withNanos}.
+  /**
+   * Sets the fraction field of a {@code TimeString} to a given number of milliseconds.
+   * This will override any fraction value previously set via {@link #withNanos}.
    *
-   * <p>For example,
-   * {@code new TimeString(1970, 1, 1, 2, 3, 4).withMillis(56)}
-   * yields {@code TIME '1970-01-01 02:03:04.056'}. */
+   * <p>For instance,
+   * {@code new TimeString("02:03:04").withMillis(56)}
+   * will produce {@code TIME '02:03:04.056'}.
+   *
+   * @param millis the milliseconds value to set;
+   *              must be between 0 (inclusive) and 1,000 (exclusive)
+   * @return a new {@code TimeString} with the specified milliseconds value
+   * @throws IllegalArgumentException if millis is outside the allowed range
+   */
   public TimeString withMillis(int millis) {
     Preconditions.checkArgument(millis >= 0 && millis < 1000);
     return withFraction(DateTimeStringUtils.pad(3, millis));
   }
 
-  /** Sets the fraction field of a {@code TimeString} to a given number
-   * of nanoseconds. Nukes the value set via {@link #withMillis(int)}.
+  /**
+   * Sets the fraction field of a {@code TimeString} to a given number of nanoseconds.
+   * This will override any fraction value previously set via {@link #withMillis(int)}.
+   *
+   * <p>For instance,
+   * {@code new TimeString("02:03:04").withNanos(56789)}
+   * will produce {@code TIME '02:03:04.000056789'}.
    *
-   * <p>For example,
-   * {@code new TimeString(1970, 1, 1, 2, 3, 4).withNanos(56789)}
-   * yields {@code TIME '1970-01-01 02:03:04.000056789'}. */
+   * @param nanos the nanoseconds value to set;
+   *             must be between 0 (inclusive) and 1,000,000,000 (exclusive)
+   * @return a new {@code TimeString} with the specified nanosecond value
+   * @throws IllegalArgumentException if nanos is outside the allowed range
+   */
   public TimeString withNanos(int nanos) {
     Preconditions.checkArgument(nanos >= 0 && nanos < 1000000000);
     return withFraction(DateTimeStringUtils.pad(9, nanos));
   }
 
-  /** Sets the fraction field of a {@code TimeString}.
-   * The precision is determined by the number of leading zeros.
-   * Trailing zeros are stripped.
+  /**
+   * Sets the fraction field of a {@code TimeString}.
+   * The precision is determined by the number of digits in the given fraction string.
+   * Trailing zeros are stripped to maintain precision.
+   *
```

Comment:
I'm not sure what 'to maintain precision' means. In a sense, when you change millis from '100' to '110' you are adding one decimal digit of precision (because the time value now has one more significant digit), and when you change it back to '100' you are losing a digit of precision.

If we really need to explain the details - and I'm not sure we do - we are maintaining the canonical form, which is that there are no trailing zeros after the decimal point.